### DISCUSSION TITLE
Allow overloading operator+= and others

### DISCUSSION THREAD

**Discussion by kaba2** [dotnet/csharplang#5795] [Language Ideas]
### Proposal

Allow overloading `operator+=` and others for
* increased performance,
* ability to avoid allocations, 
* consistency with other languages.

### Not in the proposal

It is _not_ proposed here that:
* `operator=` should be overloadable.
* operator overloads should be instance members (rather than static functions).

### How

If variable `a` has type `A` and variable `b` has type `B`, then the expression `a += b` is redefined as follows:
* If there exists a function `static C operator+=(A, B)` in `A`, then interpret `a += b` as the call expression `A.operator+=(a, b)`.
* Otherwise interpret `a += b`as `a = a + b`. 

### Notes

* The meaning of current code is preserved, because currently `operator+=` is not a valid name for a function.
* The return type `C` of `operator+=` can be arbitrary, including `void`. The most common return type is `A`.

### Example code

Example code of the proposal:

```C#
using System.Diagnostics;

int n = 1_000_000;
var v = new Vector(n);
var vBefore = v.xs;
// Interpreted as Vector.operator+=(v, v), since that functions exists.
v += v;
Debug.Assert(v.xs == vBefore);
v = v + v;
Debug.Assert(v.xs != vBefore);

// Vector in vector space R^n.
public class Vector {
    // Elements of the vector.
    public double[] xs;

    // Dimension of the vector.
    public int Dim => xs.Length;

    // Construct a zero vector of dimension `dim`.
    public Vector(int dim) {
        xs = new double[dim];
    }

    // Make a copy of this vector; memory is allocated.
    public Vector(Vector b) 
    : this(b.Dim)
    {
        xs = b.xs.ToArray();
    }

    // Add vector b to this vector without reallocations.
    // This is a mutating operation.
    public Vector Add(Vector b) {
        for (int i = 0; i < Dim; ++i) {
            xs[i] += b.xs[i];
        }
        return this;
    }

    // Current intepretation of `a += b` forces a copy.
    public static Vector operator +(Vector a, Vector b) =>
        new Vector(a).Add(b);

    // Proposed: allow to defined operator+= so that when present,
    // a += b is interpreted as a call to this function. Allocation
    // and copying is avoided.
    public static Vector operator+=(Vector a, Vector b) =>
        Add(a, b);
}
```

### Why: Current interpretation is inefficient

The current interpretation of `a += b` as  `a = a + b` leads to slow performance on vectors and matrices. Here is a simple timing for adding a million 3d-vectors together.

```C#
using System.Collections.Generic;
using System.Diagnostics;

void TimeIt(string text, Action f) {
    // Warm up.
    for (int i = 0; i < 5; ++i) f();
    var watch = Stopwatch.StartNew();
    int m = 10;
    for (int i = 0; i < m; ++i) f();
    watch.Stop();
    Console.WriteLine($"{text}: {watch.ElapsedMilliseconds / m} ms");
}

int n = 1_000_000;
int dim = 3;
var vectors = new List<Vector>(n);
for (int i = 0; i < n; ++i) {
    vectors.Add(new Vector(dim));
}

TimeIt("Current", () => AddVectors(vectors));
TimeIt("Proposed", () => AddVectorsProposed(vectors));

Vector AddVectors(List<Vector> vectors) {
    var x = new Vector(dim);
    foreach (var v in vectors) {
        x += v;
    }
    return x;
}

Vector AddVectorsProposed(List<Vector> vectors) {
    var x = new Vector(dim);
    foreach (var v in vectors) {
        // Proposed: should be able to use += instead of Add here
        // with the same performance and no allocations.
        x.Add(v);
    }
    return x;
}
```

Results on my computer:

    Current: 59 ms
    Proposed: 10 ms

Currently, the expression `x += v` creates a new vector to contain the sum `x + v`, and then assigns it to `x`. The old vector that `x` used to refer to is lost. This is inefficient. The typical 'fix' in current C# is to rewrite the addition without `+=`. I want to write in terms of operators, but still maintain efficiency.

### Why: current interpretation loses the reference on the left

There are cases in which I wish to preserve the reference on the left (e.g. having a view into a matrix column) after adding to it by `+=` . The current interpretation loses the reference 

Examples of other languages which allow overloading `operator+=` as proposed follow.

### Python

The closest example to C# is Python, because it also holds objects by reference.

```Python
class Vector:
    def __init__(self, dim):
        self.xs = [0] * dim

    def copy(self):
        v = Vector(self.dim)
        v.xs = list(v.xs)
        return v

    @property
    def dim(self):
        return len(self.xs)

    def __iadd__(a, b):
        for i in range(a.dim):
            a.xs[i] += b.xs[i]
        return a

    def __add__(a, b):
        c = a.copy()
        c += b
        return c

n = 1_000_000
v = Vector(n)
vBefore = v.xs
v += v
assert v.xs == vBefore.xs
v = v + v
assert v.xs != vBefore.xs
```

From Python's documentation:

[`object.__iadd__(self, other)`](https://docs.python.org/3/reference/datamodel.html#object.__iadd__)

> These methods are called to implement the augmented arithmetic assignments (+=, -=, *=, @=, /=, //=, %=, **=, <<=, >>=, &=, ^=, |=). These methods should attempt to do the operation in-place (modifying self) and return the result (which could be, but does not have to be, `self`). If a specific method is not defined, the augmented assignment falls back to the normal methods. For instance, if `x` is an instance of a class with an `__iadd__()` method, `x += y` is equivalent to `x = x.__iadd__(y)`. Otherwise, `x.__add__(y)` and `y.__radd__(x)` are considered, as with the evaluation of `x + y`.

Here `__iadd__` corresponds to the proposed `operator+=`, and `__add__` corresponds to `operator+`. Since I found the previous paragraph ambiguous, I'll rephrase it:

* if the function `__iadd__` exists, then `x += y` is equivalent to `x = x.__iadd__(y)`, (i.e. this is not a fallback mechanism, but how it works).
* otherwise, if `__add__` exists, then `x += y` is equivalent to `x = x.__add__(y)` (this is the first fallback mechanisms).
* otherwise, if `__radd__` exists, then `x += y` is equivalent to `x = x.__radd__(y)` (this is the second fallback mechanisms).

The proposed C# feature would work similarly to Python, with the following differences:
* Python's equivalence for `x += y` is `x = x.__iadd__(y)`. This proposal corresponds simply to `x.__iadd__(y)`. Why? Because in C# the former would force the returned type `C` and `A` to be equal for no reason. The latter is more general.
* the functions `__radd__` are not proposed.

### C++

C++ allows overloading `operator+=` as proposed. The way C++ works is closer to the proposal than Python, since `x += y` is interpreted as `Vector::operator+=(x, y)`.

[](https://en.cppreference.com/w/cpp/language/operators)

```C++
#include <algorithm>
#include <cassert>

class Vector {
public:
    double* xs;
    int dim;

    explicit Vector(int dim = 0) {
        xs = new double[dim];
        std::fill(xs, xs + dim, 0);
        this->dim = dim;
    }

    Vector(const Vector& b) 
        : Vector(b.dim)
    {
        std::copy(b.xs, b.xs + b.dim, xs);
    }

    Vector& operator=(const Vector& b) {
        std::copy(b.xs, b.xs + b.dim, xs);
        return *this;
    }

    ~Vector() {
        delete[] xs;
    }

    Vector& operator+=(const Vector& b) {
        assert(dim == b.dim);
        for (int i = 0; i < dim; ++i) {
            xs[i] += b.xs[i];
        }
        return *this;
    }

    Vector operator+(const Vector& b) const {
        return Vector(*this) += b;
    }
};

int main()
{
    int n = 1'000'000;
    auto v = Vector(n);
    double* vBefore = v.xs;
    v += v;
    assert(v.xs == vBefore);
    v = v + v;
    // In C++, assignment can also avoid reallocation
    // because operator= can be overloaded.
    assert(v.xs == vBefore);
    return 0;
}
```

### Vectors should be immutable?

From a Stack Overflow discussion on this topic:

https://stackoverflow.com/questions/6587107/c-sharp-operator-overload-for

Eric Lippert: 

> "Vectors should behave like immutable mathematical objects. When you add 2 to 3, you don't mutate the object 3 into the object 5. You create an entirely new object, 5. The point of overloading addition operators is to make your own mathematical objects; making them mutable works against that aim. I would make your vector type an immutable value type."

Since a vector a matrix can contain million elements or more, making vectors immutable is not a realistic option. Even when the vectors are just of dimension 3, allocations kill the performance of say, computing the sum of vectors as seen above. A geometric library which implements, say, B-splines (representations of curves and surfaces), makes a load of vector-computations which are naturally represented using operators. Those operations (e.g. B-spline evaluation) cannot be allowed to allocate memory to guarantee efficiency. Efficiency is of great importance to maintain real-time performance for 3D modeling programs (e.g. Blender).

**Comment by chrisoverzero** [comment]
Your benchmark has a bug, I think:

```csharp
Vector AddVectorsProposed(List<Vector> vectors) {
    var x = new Vector(dim);
    foreach (var v in vectors) {
        // Proposed: should be able to use += instead of Add here
        // with the same performance and no allocations.
        x.Add(x);
    }
    return x;
}
```

Shouldn’t that be `x.Add(v)`? Does this change the performance of this benchmark?

Additionally, you should consider using BenchmarkDotNet. There are a lot of factors which can ruin a C# benchmark, and BDN has intensive countermeasures for them.

**Reply by kaba2** [reply]
Good catch, that really should have been `x.Add(v)` (fixed it). It changes the performance to about 60ms vs 10ms. There is variation between runs, but the ratio seems roughly right. I'll try to get BenchmarkDotNet to work, I had trouble running it on my work machine.
